lib/sycamore/sycamore/writers/pinecone_writer.py

```diff
@@ -0,0 +1,207 @@
+        namespace: str,
+        dimensions: Optional[int],
+        distance_metric: str,
+        api_key: str,
+        include_binary_in_metadata: bool,
+        batch_size: int = 100,
+    ):
+        self._index_name = index_name
+        self._index_spec = index_spec
+        self._namespace = namespace
+        self._dimensions = dimensions
+        self._distance_metric = distance_metric
+        self._batch_size = batch_size
+        self._api_key = api_key
+        self._include_binary = include_binary_in_metadata
+
+    def on_write_start(self) -> None:
+        from pinecone import Pinecone
+        from pinecone.core.client.exceptions import NotFoundException
+
+        pc = Pinecone(api_key=self._api_key)
+        try:
+            pc.describe_index(self._index_name)
+        except NotFoundException:
+            if self._dimensions is None:
+                raise ValueError(f"dimensions must be specified in order to create a new index {self._index_name}")
+            if self._index_spec is None:
+                raise ValueError(f"index spec must be supplied in order to create a new index {self._index_name}")
+            pc.create_index(
+                name=self._index_name,
+                dimension=self._dimensions,
+                spec=self._index_spec,
+                metric=self._distance_metric,
+                timeout=None,
+            )
+
+    def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+        from pinecone.grpc import PineconeGRPC
+
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
+
+        pc = PineconeGRPC(api_key=self._api_key)
+        index = pc.Index(self._index_name)
+
+        objects = self._extract_pinecone_objects(block)
+        obj_it = iter(objects)
+        batch = list(itertools.islice(obj_it, self._batch_size))
+        async_results = []
+        while len(batch) > 0:
+            async_results.append(index.upsert(vectors=batch, namespace=self._namespace, async_req=True))
+            batch = list(itertools.islice(obj_it, self._batch_size))
+        for r in async_results:
+            r.result()
+
+    def _extract_pinecone_objects(self, block: Block) -> Iterable[Vector]:
+        def _add_key_to_prefix(key, prefix=""):
+            if len(prefix) == 0:
+                return str(key)
+            else:
+                return f"{prefix}.{key}"
+
+        def _flatten_metadata(data: Union[dict, list, tuple], prefix="") -> Iterable[Tuple[Any, Any]]:
+            # Pinecone requires metadata to be flat (no nested objects)
+            # so here's a traversal
+            iterator = []  # type: ignore
+            if isinstance(data, dict):
+                iterator = data.items()  # type: ignore
+            if isinstance(data, (list, tuple)):
+                iterator = enumerate(data)  # type: ignore
+            items = []
+            for k, v in iterator:
```

Comment:
Can you do an integration test where they keys are things like "\0" "a b c", "a\nb\r\n"? The pinecone docs say that keys are strings, but all the examples are /^[a-z_]+$/; we're going to have all sorts of weirdness as a result of flattening.

Reply:
```
ds.with_property("really awkward\nnam\0e", lambda doc: {"another\r\nkey": "dummy"})
```
becomes 
`properties.really awkward nam e.another key: "dummy"` (in the pinecone query ui)
You could probably commit shenanigans but I'm not too worried